services/supervisor/src/lib.rs

```diff
@@ -69,34 +98,34 @@ fn update_configs(context: &mut CallContext<'_>, changes: Vec<ConfigChange>) {
     changes.into_iter().for_each(|change| match change {
         ConfigChange::Consensus(config) => {
             log::trace!("Updating consensus configuration {:?}", config);
+            let configure_result = apply_consensus_config(context, config);
 
-            let result = context.isolate(|context| {
-                blockchain::Schema::new(context.fork())
-                    .consensus_config_entry()
-                    .set(config);
-                Ok(())
-            });
-            assert!(result.is_ok());
+            if let Err(e) = configure_result {
+                // Consensus configuration is not allowed to fail, since it was validated and approved.
+                // This is an unrecoverable error.
```

Comment:
`update_configs` is called from the `before_commit` right before the `actual_from` height.
By "unrecoverable" I mean that we've marked config as accepted already, and now we're not able to apply it, and I'm not sure what should be the right behaviour of the node in such a situation (because since it's not a transaction context, there will be no note about this failure in the blockchain).

Reply:
> there will be no note about this failure in the blockchain

Sorry for being stupid: there isn't such a note in the current implementation with a `panic`, right? Do we need one? I *think* it could be possible to design it, e.g., by using a flag:

```rust
Schema::new(context.fork()).set_flag(false); // is the call successful?
context.isolate(|context| {
    potentially_panic(context)?;
    Schema::new(context.fork()).set_flag(true);
    Ok(())
})
```

...but I'm not sure we need this for consensus config at least (since I'm not sure that a panic can occur there). For services it may make more sense, since the failure path is obvious.

Reply:
Yep, you're right. If transaction panics, this is considered a transaction failure, so we can see the note about `ErrorKind::Panic` inside the blockchain. If service panics during `before_commit`, the rollback happens and... that's it: [1](https://github.com/exonum/exonum/blob/dynamic_services/exonum/src/runtime/rust/mod.rs#L388), [2](https://github.com/exonum/exonum/blob/dynamic_services/exonum/src/runtime/dispatcher/mod.rs#L265) (I don't consider adding a log entry as a valid network-wide note). So, we aren't able to distinguish a situation when service did nothing during `before_commit` from the situation when service panicked.

Do we need one? I'm not sure until we discuss what policy on panics do we have overall.
Current state "panics are caught and suppressed" is not covering the state of the blockchain after the accident, which may be somewhat "poisoned" as in this particular case.

So I think this topic deserves a discussion (bigger than conversation in the PR comments) and maybe some r&d before we will be able to answer this question.